x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -649,142 +711,270 @@ private void indexInvalidation(Collection<String> tokenIds, ActionListener<Token
-                                        e -> {
-                                            Throwable cause = ExceptionsHelper.unwrapCause(e);
-                                            if (cause instanceof VersionConflictEngineException ||
-                                                isShardNotAvailableException(e)) {
-                                                innerRefresh(tokenDocId, clientAuth,
-                                                    listener, attemptCount);
-                                            } else {
-                                                onFailure.accept(e);
-                                            }
-                                        }),
-                                    client::update);
-                            }
+                            logger.warn("could not find superseding token document [{}] for token document [{}] after all retries",
+                                supersedingTokenDocId, tokenDocId);
+                            onFailure.accept(invalidGrantException("could not refresh the requested token"));
                         }
-                    } else {
-                        logger.info("could not find token document [{}] for refresh", tokenDocId);
-                        onFailure.accept(invalidGrantException("could not refresh the requested token"));
                     }
-                }, e -> {
-                    if (isShardNotAvailableException(e)) {
-                        innerRefresh(tokenDocId, clientAuth, listener, attemptCount);
-                    } else {
-                        listener.onFailure(e);
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        if (isShardNotAvailableException(e)) {
+                            if (backoff.hasNext()) {
+                                logger.info("could not find superseding token document [{}] for refresh, retrying", supersedingTokenDocId);
+                                client.threadPool().schedule(
+                                    () -> getTokenDocAsync(supersedingTokenDocId, this), backoff.next(), GENERIC);
+                            } else {
+                                logger.warn("could not find token document [{}] for refresh after all retries", supersedingTokenDocId);
+                                onFailure.accept(invalidGrantException("could not refresh the requested token"));
+                            }
+                        } else {
+                            logger.warn("could not find superseding token document [{}] for refresh", supersedingTokenDocId);
+                            onFailure.accept(invalidGrantException("could not refresh the requested token"));
+                        }
                     }
-                }), client::get);
+                };
+                getTokenDocAsync(supersedingTokenDocId, getSupersedingListener);
+            } else {
```

Comment:
Here, it is not "eligible for multiple refresh". So, we assume it is not refreshed at all. I believe it might as well be that it was refreshed a long time ago?

Reply:
> , we assume it is not refreshed at all

Not exactly. We first check if it is eligible for refresh in `checkTokenDocForRefresh` so we know it was issued in the last 24 hrs - then we check if it is eligible for multi refresh i.e. issued in the last seconds.

Reply:
You're right, I got lost in it.